Updates query DSL index page (#2451)
opensearch-project/documentation-website: _opensearch/query-dsl/index.md

# Query DSL

OpenSearch provides a query domain-specific language (DSL) that you can use to perform customized searches on your data.

> Query DSL provides several query categories. You can even perform more complicated searches by using different elements from each variety to find more granular data that you need.

Author: thanks. I'll clarify it a bit. they can use multiple fields for more granular searches.

The following query DSL categories include:

> - **Compound boolean queries** – To search with clause rules that act as logical operators to combine several simple query terms.

Reviewer: Even though we don't have all of the documentation yet, should we list all query types here and not only the ones for which we do have documentation?

Reviewer 2: I agree. Let's provide a list of the ones that we support. We can always link to the documentation for each one later.

Author: We would need those sections to be available on the site. This is a tech doc, not marketing "what's coming" announcement.

Author: ok will do. 
The reason I didn't list them here is because it would be confusing to see it called out in the overview, and then customers can't go to the missing section.

> - **Term-level queries** – To search documents for one or more terms, IDs, value ranges, with optional filtering by wildcard or regex, but does not provide any relevance score or text analysis.

Reviewer: Term-level queries calculate the relevance score, but they don't sort by relevance score. So maybe
```suggestion
- **Term-level queries** – To search documents for one or more terms, IDs, value ranges, with optional filtering by wildcard or regex. Term-level queries do not sort results by relevance score or analyze search terms.
```

- **Full-text queries** – To search documents for one or more terms and filter with advanced options. These queries return detailed information including relevance scores per document match, and perform text analysis.

> - **Geo-bounding box queries** – To search documents within a range of `geo_point` field values.

Reviewer: ```suggestion
- **Geo-bounding box queries** – To search for documents with `geo_point` field values that are within a bounding box.
```

> - **xy point queries** – To search two-dimensional coordinates with `xy_point` or `xy_shape` fields.

Reviewer: ```suggestion
- **xy point queries** – To search for documents with two-dimensional coordinates in `xy_point` or `xy_shape` fields.
```

# Query DSL

OpenSearch provides a query domain-specific language (DSL) that you can use to perform customized searches on your data. You can also combine query fields to perform more complex searches and find more granular data.

OpenSearch supports the following query DSL query type categories:

- **Compound queries** – Used to perform combined queries with the following query types:
    - **Boolean** `bool` – Combines several simple query terms by Boolean clause rules that act as logical operators. To learn more, see [Boolean queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/bool/).

>     - **Constant score** `constant_score` – Provides a `filter` query with the relevance score as a floating point number assigned by the `boost` field. Acts as a wrapper for the filter.

Reviewer: ```suggestion
    - **Constant score** `constant_score` – Uses a `filter` query to return all matching documents and gives each document the same relevance score that is equal to the `boost` value. 
```

>     - **Disjunction max** `dis_max` – Returns documents that match one or more queries that are wrapped together by clauses. If a document matches multiple query clauses, it gets the highest relevant score assigned to it.

Reviewer: ```suggestion
    - **Disjunction max** `dis_max` – Returns documents that match one or more query clauses. If a document matches multiple query clauses, its relevance score is calculated using the highest score from any matching clause and, optionally, the scores from the other matching clauses multiplied by the tie breaker value.
```

    - **Function score** `function_score` – Modifies the score of documents that are returned by a query. You define a query and one or more functions to compute a score for each document that matches the query.

> - **Full-text queries** – Used to search documents for one or more terms and filter with advanced options. These queries perform text analysis and return detailed information, including relevance scores for each document match. To learn more, see [Full-text queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/full-text/).

Reviewer: If we're listing all the subtypes of the compound query, should we list all the subtypes of the full text query here?

Author: Because the compound query types listed here do not have topic pages we can link to (such as full-text), they need to be included in the index bullet list of descriptions.

The goal is to have a concise bullet list of category descriptions and link to their topic pages (if available).

Author: Other option is to create a sub-page for compound queries that would only contain these five type definitions.

> - **Geographic and cartesian shape queries** – To search documents that include geographic geopoint field data, or to search two-dimensional geometries that map out cartesian data such as the `xy_point` field that supports x and y pairs and the `xy_shape` field that supports points, lines, circles and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).

Reviewer 3: "Used to search documents that include geographic geopoint field data or two-dimensional geometries that map out cartesian data, such as the `xy_point` field that supports x and y pairs and the `xy_shape` field that supports points, lines, circles, and polygon shapes." ?

Lots of "that"s in this sentence. Might break it up into two.

> - **Geographic and cartesian shape queries** – Used to search documents that include geographic geopoint field data. Also used to search two-dimensional geometries that map out cartesian data, such as the `xy_point` field that supports x and y pairs and the `xy_shape` field that supports points, lines, circles, and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).

Reviewer 4: ```suggestion
- **Geographic and cartesian shape queries** – Used to search documents that include geographic geopoint field data. Also used to search two-dimensional geometries that map out cartesian data, such as the `xy_point` field, which supports x and y pairs, and the `xy_shape` field, which supports points, lines, circles, and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).
```

Reviewer 4: Is "geographic geopoint" redundant?

Reviewer: You can search for documents, which contain points and shapes in a Cartesian coordinate system using xy queries. To search for documents, which contain geographic points and shapes, use geographic queries.
```suggestion
- **Geographic and cartesian shape queries** – Use geographic queries to search documents that include geographic data. Use xy queries to search documents that include points and shapes in a two-dimensional Cartesian coordinate system. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).
```

> - **Joining queries** – Used to perform `nested`, `has_child`, `has_parent`, and `parent_id` queries.

Reviewer: ```suggestion
- **Joining queries** – Used to search nested fields or return parent and child documents that match a specific query. Types of joining queries include `nested`, `has_child`, `has_parent`, and `parent_id` queries.
```

> - **Query string queries** – Used to perform complex queries for a string with multiple optional fields using `query_string` syntax. To learn more, see [Query string queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/query-string/).

Reviewer 4: The first sentence here is a bit confusingly written. Can we simplify/clarify?

Reviewer: Query string query should go under "full text queries". It's a subtype of full text.
```suggestion
- **Query string queries** – Query string queries use a more concise `query_string` syntax. They can incorporate wildcards and search multiple fields. To learn more, see [Query string queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/query-string/).
```

> - **Span queries** – Used to perform queries that provide control over the order and proximity of query terms that you specify. The primary use case is for legal documents.  To learn more, see [Span queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/span-query/).

Reviewer 4: ```suggestion
- **Span queries** – Used to perform queries that provide control over the order and proximity of specified query terms. Primarily used for legal documents. To learn more, see [Span queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/span-query/).
```

Reviewer 4: Global: Ensure one space following a period throughout.

> - **Span queries** – Used to perform queries that provide control over the order and proximity of query terms that you specify. The primary use case is for legal documents.  To learn more, see [Span queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/span-query/).

Reviewer: Same as above: should we list the types of span queries here?
```suggestion
- **Span queries** – Span queries provide control over the order and proximity of query terms that you specify. The primary use case is to create highly specific queries for legal documents.  To learn more, see [Span queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/span-query/).
```

Author: we link to a sub-page to make the index bullet list concise.

> - **Specialized queries** – Used to perform specialized queries: `distance_feature`, `more_like_this`, `percolate`, `rank_feature`, `script`, `script_score`, `wrapper`, or `pinned_query`.

Reviewer: ```suggestion
- **Specialized queries** – Specialized queries include the following query categories: `distance_feature`, `more_like_this`, `percolate`, `rank_feature`, `script`, `script_score`, `wrapper`, or `pinned_query`.
```

>  - **Term-level queries** – Used to search documents for one or more terms, IDs, value ranges, with optional filtering by wildcard or regex. Term-level queries do not sort results by relevance score or analyze search terms. To learn more, see [Term-level queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/term/).

Reviewer 4: ```suggestion
 - **Term-level queries** – Used to search documents for one or more terms, IDs, or value ranges, with optional filtering by wildcard or regex. Term-level queries do not sort results by relevance score or analyze search terms. To learn more, see [Term-level queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/term/).
```

## About term-level and full-text queries

> Although term-level and full-text queries both search for terms, there are some differences as described in the following table:

Reviewer 4: ```suggestion
Although term-level and full-text queries both search for terms, there are some differences between the two, as described in the following table.
```

    - **Function score** `function_score` – Modifies the score of documents that are returned by a query. You define a query and one or more functions that each compute a score for each document that matches the query.

- **Full-text queries** – To search documents for one or more terms and filter with advanced options. These queries return detailed information including relevance scores per document match, and perform text analysis. To learn more, see [Full-text queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/full-text/).

- **Geographic and cartesian shape queries** – To search documents that include geographic geopoint field data, or to search two-dimensional geometries that map out cartesian data such as the `xy_point` field that supports x and y pairs and the `xy_shape` field that supports points, lines, circles and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).

- **Joining queries** – To perform `nested`, `has_child`, `has_parent`, and `parent_id` queries.

> - **Query context** – Indicates Boolean query clauses to indicate specifically how well a document matches a query based on the `relevance_score` returned with a query.

Reviewer: Query and filter contexts are not types of query DSL queries. I would remove them from this list.

# Query DSL

> OpenSearch provides a query domain-specific language (DSL) that you can use to perform customized searches on your data.You can even perform more complicated searches by combining query fields to find more granular data that you need.

Reviewer: ```suggestion
OpenSearch provides a query domain-specific language (DSL) that you can use to perform customized searches on your data. You can even perform more complicated searches by combining query fields to find more granular data that you need.
```

    - **Boolean** `bool` – Combines several simple query terms by boolean clause rules that act as logical operators. To learn more, see [Boolean queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/bool/).
    - **Constant score** `constant_score` – Provides a `filter` query with the relevance score as a floating point number assigned by the `boost` field. Acts as a wrapper for the filter.
    - **Disjunction max** `dis_max` – Returns documents that match one or more queries that are wrapped together by clauses. If a document matches multiple query clauses, it gets the highest relevant score assigned to it.
    - **Function score** `function_score` – Modifies the score of documents that are returned by a query. You define a query and one or more functions that each compute a score for each document that matches the query.

- **Full-text queries** – To search documents for one or more terms and filter with advanced options. These queries return detailed information including relevance scores per document match, and perform text analysis. To learn more, see [Full-text queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/full-text/).

> - **Geographic and shape queries** – To create queries to search two-dimensional geometries that map out cartesian data such as `point` fields that support x and y pairs and `shape` fields that support points, lines, circles and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).

Reviewer: In OpenSearch, we don't have a `point` and `shape` field types. We have an `xy point` and xy shape` field types.
```suggestion
- **Geographic and shape queries** – To create queries to search two-dimensional geometries that map out cartesian data such as `xy point` fields that support x and y pairs and `xy shape` fields that support points, lines, circles and polygon shapes. To learn more, see [Geographic and shape queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/geo-and-shape/).
```

- **Joining queries** – To perform `nested`, `has_child`, `has_parent`, and `parent_id` queries.

> - **Parameter queries** – To perform `minimum_should_match` and `rewrite` queries.

Reviewer: There are no "Parameter queries" as such. `minimum_should_match` and `rewrite` are parameters that can be used in queries. We can either talk about the parameters in a different list, or remove this line altogether.

> - **Query context** – To use the default `relevance_score` returned with a query to indicate how well a document matches a query, or whether or not it matches at all with the following query clauses:

Reviewer: Same comment as above. We can either talk about the query and filter contexts in a different list, or remove this line altogether.

    }
```

### Using query DSL to create customized searches

> With query DSL you can include an HTTP request body to look for results more tailored to your needs. The following example shows how to search for `speaker` and `text_entry` fields that have a value of `QUEEN`:

Reviewer 3: "... better tailored to your needs." ?
